src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs

```diff
@@ -61,63 +60,90 @@ internal override BoundStatement BindUsingStatementParts(DiagnosticBag diagnosti
 
-                Debug.Assert(!declarations.IsEmpty);
+                    Debug.Assert(!declarations.IsEmpty);
 
-                declarationsOpt = new BoundMultipleLocalDeclarations(declarationSyntax, declarations);
+                    declarationsOpt = new BoundMultipleLocalDeclarations(declarationSyntax, declarations);
 
-                TypeSymbol declType = declarations[0].DeclaredType.Type;
+                    TypeSymbol declType = declarations[0].DeclaredType.Type;
 
-                if (declType.IsDynamic())
-                {
-                    iDisposableConversion = Conversion.ImplicitDynamic;
-                }
-                else
-                {
-                    HashSet<DiagnosticInfo> useSiteDiagnostics = null;
-                    iDisposableConversion = originalBinder.Conversions.ClassifyImplicitConversionFromType(declType, iDisposable, ref useSiteDiagnostics);
-                    diagnostics.Add(declarationSyntax, useSiteDiagnostics);
-
-                    if (!iDisposableConversion.IsImplicit)
+                    if (declType.IsDynamic())
                     {
-                        if (!declType.IsErrorType())
+                        iDisposableConversion = Conversion.ImplicitDynamic;
+                    }
+                    else
+                    {
+                        HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                        iDisposableConversion = originalBinder.Conversions.ClassifyImplicitConversionFromType(declType, iDisposable, ref useSiteDiagnostics);
+                        diagnostics.Add(declarationSyntax, useSiteDiagnostics);
+
+                        if (!iDisposableConversion.IsImplicit)
                         {
-                            Error(diagnostics, ErrorCode.ERR_NoConvToIDisp, declarationSyntax, declType);
+                            if (!declType.IsErrorType())
+                            {
+                                Error(diagnostics, hasAwait ? ErrorCode.ERR_NoConvToIAsyncDisp : ErrorCode.ERR_NoConvToIDisp, declarationSyntax, declType);
+                            }
+
+                            hasErrors = true;
                         }
+                    }
+                }
 
-                        hasErrors = true;
+                if (!hasErrors && hasAwait)
+                {
+                    TypeSymbol taskType = this.Compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task);
+                    hasErrors = ReportUseSiteDiagnostics(taskType, diagnostics, _syntax.AwaitKeyword.GetLocation());
+
+                    if (!hasErrors)
+                    {
+                        var syntaxForAwait = (SyntaxNode)expressionSyntax ?? declarationSyntax;
+                        BoundExpression placeholder = new BoundAwaitableValuePlaceholder(syntaxForAwait, taskType);
+                        boundAwaitOpt = BindAwait(placeholder, syntaxForAwait, diagnostics).MakeCompilerGenerated();
```

Comment:
We're marking the `BoundAwaitExpression` as compiler generated here, but not in existing `await` scenarios. Why the difference?

Reply:
> We're marking the BoundAwaitExpression as compiler generated here, but not in existing await scenarios. Why the difference?

I guess because there is no corresponding AwaitExpressionSyntax.

---
In reply to: [160518475](https://github.com/dotnet/roslyn/pull/23961#discussion_r160518475) [](ancestors = 160518475)

Reply:
Because in a case of Using, the await expression is not isomorphic to any expression in the syntax.
We are using a bound node just as a storage for method/property symbols found in binding. The node has no other structure or nesting (like in deconstruct).  
I think `AwaitableInfo`, which is not a bound node, would make things simpler.

Reply:
The main reason why I marked it as compiler-generated is that `AsyncMethodToStateMachineRewriter.VisitAwaitExpression` expects that a `BoundAwaitExpression` is attached to some `await` syntax.
For async-using, we have no `await` syntax, so I modified the assertion as shown below.

Also, I think it makes sense into terms of bound tree having the same shape as the syntax tree.

```C#
        private BoundBlock VisitAwaitExpression(BoundAwaitExpression node, BoundExpression resultPlace)
        {
            ...
            Debug.Assert(node.Syntax.IsKind(SyntaxKind.AwaitExpression) || node.WasCompilerGenerated);
```

---
In reply to: [160567590](https://github.com/dotnet/roslyn/pull/23961#discussion_r160567590) [](ancestors = 160567590)